### PR TITLE
Fix segfault on zero-length slice assignment on memoryview

### LINKED ISSUE #6227: [BUG] Segfault on zero-length slice assignment on memoryview
### Describe the bug

We've experienced an issue in production where attempting to write a `folly:IOBuf` with zero-length backing buffers is triggering a SIGSEGV. 

We're traced this issue to the way Cython is executing zero-length slices on `cmemoryview`, e.g. `view[0:0]`. While this problem was trivially addressed by no-oping zero-length writes to our stream, this behavior represents a divergence from the standard Python behavior where zero-length slice access works as intended.

### Code to reproduce the behaviour:

main.pyx: Create a struct that wraps a Python `memoryview` and exposes a function to write nothing to an empty slice
```cython
from cython.view cimport memoryview as cmemoryview
from cpython.buffer cimport PyBUF_C_CONTIGUOUS

cdef class MemoryViewStream:
  cdef cmemoryview view

  def __cinit__(self, object view):
    self.view = cmemoryview(view, PyBUF_C_CONTIGUOUS)

  def write_nothing(self):
    self.view[0:0] = []
```

test.py
```python
from main import MemoryViewStream

# empty byte array of len 10
dest_byte_arr = bytearray(10)
mem_view = memoryview(dest_byte_arr)

# create C struct to represent this stream
stream = MemoryViewStream(mem_view)

# write zero length bytearray to the stream
# Segfaults here
stream.write_nothing()
print("Completed successfully")
```

setup.py
```python
from setuptools import setup
from Cython.Build import cythonize

setup(
    ext_modules = cythonize("main.pyx")
)
```

### Expected behaviour

We expect `python test.py` to execute successfully and print "Completed successfully" — however, instead we end up with a SIGSEGV due to an invalid memory access in the C stdlib.

### OS

Repro'd on both Mac and CentOS

### Python version

3.9.6

### Cython version

3.0.10

### Additional context

Stack trace on SIGSEGV indicates that C `strlen` is being invoked on a null pointer, and most likely a fix is required in `__pyx_memoryview_assign_item_from_object`

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	       0x184d18904 _platform_strlen + 4
1   Python3                       	       0x104c75e2c PyBytes_FromString + 24
2   main.cpython-39-darwin.so     	       0x10445a95c __pyx_memoryview_assign_item_from_object + 132 (main.c:8673)
3   main.cpython-39-darwin.so     	       0x104459e90 __pyx_memoryview_setitem_slice_assign_scalar + 172 (main.c:8149)
4   main.cpython-39-darwin.so     	       0x104464c50 __pyx_memoryview___pyx_pf_15View_dot_MemoryView_10memoryview_6__setitem__ + 1044 (main.c:7623) [inlined]
5   main.cpython-39-darwin.so     	       0x104464c50 __pyx_memoryview___setitem__ + 1044 (main.c:7469) [inlined]
6   main.cpython-39-darwin.so     	       0x104464c50 __pyx_mp_ass_subscript_memoryview + 1304 (main.c:18513)
7   main.cpython-39-darwin.so     	       0x10445b5c4 __Pyx_PyObject_SetSlice + 36 (main.c:22955) [inlined]
8   main.cpython-39-darwin.so     	       0x10445b5c4 __pyx_pf_4main_16MemoryViewStream_2write_nothing + 56 (main.c:17573) [inlined]
9   main.cpython-39-darwin.so     	       0x10445b5c4 __pyx_pw_4main_16MemoryViewStream_3write_nothing + 88 (main.c:17550)
10  Python3                       	       0x104d52a38 0x104c40000 + 1124920
11  Python3                       	       0x104d4fb54 _PyEval_EvalFrameDefault + 22664
12  Python3                       	       0x104d537ec 0x104c40000 + 1128428
13  Python3                       	       0x104d4a204 PyEval_EvalCode + 80
14  Python3                       	       0x104d8e2f0 0x104c40000 + 1368816
15  Python3                       	       0x104d8e4c4 0x104c40000 + 1369284
16  Python3                       	       0x104d8c5ec PyRun_SimpleFileExFlags + 820
17  Python3                       	       0x104da97cc Py_RunMain + 1604
18  Python3                       	       0x104da9c3c 0x104c40000 + 1481788
19  Python3                       	       0x104da9cdc Py_BytesMain + 40
20  dyld                          	       0x1849620e0 start + 2360

### PR DESCRIPTION
**Pull request by michael99man**
As outlined in #6227 — we have experienced an issue in production where attempting to write a folly:IOBuf with zero-length backing buffers is triggering a SIGSEGV.

The specific issue actual has to do with how `memoryview` attempts to use scalar assignment instead of slice assignment in the case where the source buffer (in this case, an empty slice) has `len == 0`. By fixing the fork to properly use slice assignment, we avoid the SIGSEGV when we string format the view format (which is NULL).

This PR also adds two additional tests for writing either a single byte or an empty slice to a `memoryview`.

Tested via: `python runtests.py memslice -vv` (which includes the two new tests)

fixes #6227

### CODE REVIEW

File: Cython/Utility/MemoryView.pyx [hunk 1]
```diff
@@ -417,7 +417,7 @@ cdef class memoryview:
 
         if have_slices:
             obj = self.is_slice(value)
-            if obj:
+            if obj != None:
```
[the remark this answers is not in the document]
Comment (michael99man) [reply Cython/Utility/MemoryView.pyx]: Good catch — got tripped up between C/Python :P

File: tests/memoryview/memslice.pyx [hunk 2]
```diff
@@ -2177,6 +2180,40 @@ cdef _test_slice_assignment_broadcast_strides(slice_1d src, slice_2d dst, slice_
...
+    >>> test_slice_assignment_zero_length_slice()
+    0
+    """
+    cdef char[1] src_buf
+    src_buf[0] = 0x01
+
+    dest_buf: bytearray = bytearray(1)
+
+    # construct a memory view of the destination buffer
+    view = memoryview(dest_buf, PyBUF_C_CONTIGUOUS)
```
[the remark this answers is not in the document]
Comment (michael99man) [reply tests/memoryview/memslice.pyx:2209]: Under the hood, does this do the same memory mapping between the Python bytearray and the memoryview object?

**Comment by michael99man** [comment]
Thanks for taking a look! Applied suggestions :)

**Comment by michael99man** [comment]
Seems like the test changes broke the tests due to `TypeError: an integer is required` — I refactored the tests via the cdef helper pattern used by other tests in this file, and tests seem to be working again

**Comment by scoder** [comment]
I don't think this is the right fix. Here's a test that fails for me and succeeds if I make your change in `MemoryView.pyx`:
```patch
diff --git a/tests/memoryview/memoryview.pyx b/tests/memoryview/memoryview.pyx
index de5fa4f21..5f95559c0 100644
--- a/tests/memoryview/memoryview.pyx
+++ b/tests/memoryview/memoryview.pyx
@@ -1081,6 +1081,22 @@ def test_dtype_object_scalar_assignment():
     assert m[0] == m[4] == m[-1] == 3
 
 
+def test_assign_to_slice(obj, start, end):
+    """
+    >>> test_assign_to_slice(b'abc', 0, 3)
+    b'abc'
+    >>> test_assign_to_slice(b'a', 0, 1)
+    b'a'
+    >>> test_assign_to_slice(b'', 0, 0)
+    b''
+    >>> test_assign_to_slice(b'', 5, 5)
+    b''
+    """
+    view = memoryview(bytearray(len(obj)), PyBUF_C_CONTIGUOUS)
+    view[start:end] = obj[start:end]
+    return bytes(view)
+
+
 def test_assignment_in_conditional_expression(bint left):
     """
     >>> test_assignment_in_conditional_expression(True)
@@ -1310,5 +1326,5 @@ def test_untyped_index(i):
 
 _PERFORMANCE_HINTS = """
 243:9: Use boundscheck(False) for faster access
-1313:21: Index should be typed for more efficient access
+1325:21: Index should be typed for more efficient access
 """
```

Sorry for the wrong changes to your test. I got confused by the fact that you put it into `memslice.pyx` rather than `memoryview.pyx`, but looking at your original bug report, it needs to go there since it's a test for the actual `memoryview` Python type.

**Comment by michael99man** [comment]
Good callout, my "fixed" tests didn't actually repro the segfault. 

I thought to put the test in `memoryview.pyx` but got confused by the "slice" terminology 😅 — thanks for providing an updated test! I've also verified that this new test works with the fix and segfaults without it.

File: tests/memoryview/memoryview.pyx [hunk 3]
```diff
@@ -1314,5 +1330,5 @@ def test_untyped_index(i):
 
 _PERFORMANCE_HINTS = """
 243:9: Use boundscheck(False) for faster access
-1313:21: Index should be typed for more efficient access
+1329:21: Index should be typed for more efficient access
```
Comment (michael99man) [inline tests/memoryview/memoryview.pyx:1333]: This performance hint was different from your example (1329 vs. 1325), perhaps due to code staleness?

Comment (da-woods) [reply tests/memoryview/memoryview.pyx:1333]: It's just a line number

Comment (michael99man) [reply tests/memoryview/memoryview.pyx:1333]: Yeah, I'm not clear on why @scoder's line number differs from mine. I think L1329 is the correct line based off this current revision

Comment (scoder) [reply tests/memoryview/memoryview.pyx:1333]: Might have been an incomplete manual change on my part. If CI is happy, I'll be happy.

**Comment by scoder** [comment]
Thanks.

Does anyone care about a backport? Seems appropriate to me. The test won't apply trivially to the 3.0.x branch because the doctest output was written for Python 3.x. And performance hints weren't there yet, so that part needs to be removed.
